src/lib/comms/sol-lwm2m.c

```diff
@@ -2822,19 +2822,25 @@ handle_read(struct sol_lwm2m_client *client,
         }
     }
 
-    SOL_VECTOR_FOREACH_IDX (&resources, res, i) {
-        r = setup_tlv(res, &buf);
-        SOL_INT_CHECK_GOTO(r, < 0, err_exit);
-        sol_lwm2m_resource_clear(res);
-    }
-
     r = add_coap_int_option(resp, SOL_COAP_OPTION_CONTENT_FORMAT,
         &format, sizeof(format));
     SOL_INT_CHECK_GOTO(r, < 0, err_exit);
 
     r = set_packet_payload(resp, (const uint8_t *)buf.data, buf.used);
     SOL_INT_CHECK_GOTO(r, < 0, err_exit);
 
+    SOL_VECTOR_FOREACH_IDX (&resources, res, i) {
+        r = setup_tlv(res, &buf);
+        SOL_INT_CHECK_GOTO(r, < 0, err_exit);
+    }
```

Comment:
I'm not sure we can change the order here, @cabelitos or someone needs to check the rest of the code, but usually setting coap payload depends on all headers to be set as it will follow from that position (encodes to the wire). Then if setup_tlv() does headers, it maybe of problems. Another possibility of problem is that `set_packet_payload()` uses buffer, that is set by `setup_tlv()`.